### PR TITLE
Osipov/unit tests

### PR DESCRIPTION
**Pull request by zouxuoz**

### CODE REVIEW

File: client/lib/bem/bem.js [hunk 1]
```diff
@@ -7,15 +7,19 @@ const bem = block => (element, modifiers = {}) => {
...
 
-  const modifierClassNames = Object.keys(modifiers).map((modifier) => {
-    const value = modifiers[modifier];
+  const modifierClassNames = Object.keys(modifiers).reduce((result, modifierName) => {
+    const value = modifiers[modifierName];
 
     if (typeof value === 'boolean') {
-      return `${rootClassName}_${modifier}`;
+      if (value) {
+        result = result.concat(`${rootClassName}_${modifierName}`);
```
Comment (rosendi) [inline client/lib/bem/bem.js]: Just return `result.concat...`

File: client/lib/bem/bem.js [hunk 2]
```diff
@@ -7,15 +7,19 @@ const bem = block => (element, modifiers = {}) => {
...
+      }
+    } else {
```
Comment (rosendi) [inline client/lib/bem/bem.js]: Remove `else`

File: client/lib/bem/bem.js [hunk 3]
```diff
@@ -7,15 +7,19 @@ const bem = block => (element, modifiers = {}) => {
...
+      result = result.concat(`${rootClassName}_${modifierName}_${value}`);
```
Comment (rosendi) [inline client/lib/bem/bem.js]: Just return `result.concat...`

File: client/lib/bem/bem.js [hunk 4]
```diff
@@ -7,15 +7,19 @@ const bem = block => (element, modifiers = {}) => {
...
     }
 
-    return `${rootClassName}_${modifier}_${value}`;
-  });
+    return result;
```
Comment (rosendi) [inline client/lib/bem/bem.js]: Remove

File: client/logic/modules/auth/__tests__/reducers.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
+import { AUTHENTICATE, FETCH_USERDATA_REQUEST, FETCH_USERDATA_SUCCESS, FETCH_USERDATA_FAILURE, LOGOUT_REQUEST, LOGOUT_SUCCESS, LOGOUT_FAILURE } from '../actions';
+import authReducer from '../reducers';
+
+test('should return the initial state', () => {
+  const expectedState = {
+    token: null,
```
Comment (rosendi) [inline client/logic/modules/auth/__tests__/reducers.js]: put \n and in the section below.

File: client/logic/modules/auth/__tests__/reducers.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+test('should handle AUTHENTICATE', () => {
+  const expectedState = {
+    token: 'token',
+    userdata: {
+      fetching: false,
+    },
+  };
+
+  expect(authReducer({}, {
```
Comment (rosendi) [inline client/logic/modules/auth/__tests__/reducers.js]: I really think we can write this construction in a one line if they are simple like here.

File: client/logic/modules/auth/sagas.js [hunk 7]
```diff
@@ -5,7 +5,7 @@ import { AUTHENTICATE, LOGOUT_REQUEST, LOGOUT_SUCCESS, FETCH_USERDATA_REQUEST }
 import { authenticate, fetchUserdata, fetchUserdataDone, fetchUserdataError, logoutDone, logoutError } from './creators';
 import { getAuthToken, setAuthToken, purgeAuthToken } from './services';
 
-function* auth() {
+export function* auth() {
```
Comment (rosendi) [inline client/logic/modules/auth/sagas.js]: Well it's good question. In general if a saga is private you don't have to test it. That saga is private. Tomorrow I'l check it.

File: package.json [hunk 8]
```diff
@@ -3,6 +3,9 @@
   "version": "2.0.0-beta.6",
   "private": true,
   "scripts": {
+    "test": "jest",
+    "test:watch": "npm test -- --watch",
```
Comment (rosendi) [inline package.json]: Yeah! It's really cool!

File: package.json [hunk 9]
```diff
@@ -108,5 +105,10 @@
   },
   "engines": {
     "npm": ">=2.0.0"
+  },
+  "jest": {
+    "rootDir": "",
+    "testRegex": "/__tests__/",
```
Comment (rosendi) [inline package.json]: It's default, isn't it? :-)

File: client/lib/bem/__tests__/bem.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+});
+
+test('should return element className for element', () => {
+  const className = bem('block')('element').className;
+
+  expect(className).toBe('block__element');
+});
+
+test('should return element className with boolean modifiers for element with boolean modifiers', () => {
+  const className = bem('block')('element', { foo: true, hoo: false }).className;
```
Comment (zouxuoz) [inline client/lib/bem/__tests__/bem.js]: fixed

File: client/lib/bem/bem.js (hunk 1 above)
Comment (zouxuoz) [inline client/lib/bem/bem.js]: fixed

File: client/lib/bem/bem.js (hunk 2 above)
Comment (zouxuoz) [inline client/lib/bem/bem.js]: fixed

File: client/lib/bem/bem.js (hunk 3 above)
Comment (zouxuoz) [inline client/lib/bem/bem.js]: fixed

File: client/logic/modules/auth/__tests__/reducers.js (hunk 5 above)
Comment (zouxuoz) [inline client/logic/modules/auth/__tests__/reducers.js]: fixed

File: client/logic/modules/auth/__tests__/reducers.js (hunk 6 above)
Comment (zouxuoz) [inline client/logic/modules/auth/__tests__/reducers.js]: fixed

File: package.json (hunk 9 above)
Comment (zouxuoz) [inline package.json]: fixed

File: client/logic/modules/auth/__tests__/creators.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+test('should create an action for logout error', () => {
+  const expectedAction = {
+    type: LOGOUT_FAILURE,
+    reason: 'reason',
+  };
+
+  expect(logoutError('reason')).toEqual(expectedAction);
+});
+
+test('should create an action for fetch user data', () => {
```
Comment (rosendi) [inline client/logic/modules/auth/__tests__/creators.js]: creates  an action to fetch user data

File: client/logic/modules/auth/__tests__/creators.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  const expectedAction = {
+    type: FETCH_USERDATA_REQUEST,
+  };
+
+  expect(fetchUserdata()).toEqual(expectedAction);
+});
+
+test('should create an action for fetch user data done', () => {
```
Comment (rosendi) [inline client/logic/modules/auth/__tests__/creators.js]: creates FETCH_USERDATA_SUCCESS when fetching user data has been done

File: client/logic/modules/auth/__tests__/creators.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  const expectedAction = {
+    type: FETCH_USERDATA_SUCCESS,
+    payload: 'payload',
+  };
+
+  expect(fetchUserdataDone('payload')).toEqual(expectedAction);
+});
+
+test('should create an action for fetch user data error', () => {
```
Comment (rosendi) [inline client/logic/modules/auth/__tests__/creators.js]: creates FETCH_USERDATA_SUCCESS when fetching user data has been failed

File: .gitignore [hunk 14]
```diff
@@ -11,3 +11,4 @@
 /test/e2e/target
 /translations
 npm-debug.log
+coverage
```
Comment (rosendi) [inline .gitignore]: /coverage and sort alpha-y

File: client/logic/modules/auth/reducers.js [hunk 15]
```diff
@@ -1,13 +1,13 @@
...
+import { AUTHENTICATE, LOGOUT_SUCCESS, FETCH_USERDATA_REQUEST, FETCH_USERDATA_SUCCESS, FETCH_USERDATA_FAILURE } from './actions';
 
 const token = (state = null, { type, payload }) => {
   switch (type) {
     case AUTHENTICATE: {
       return payload;
     }
 
-    case LOGOUT: {
+    case LOGOUT_SUCCESS: {
```
Comment (rosendi) [inline client/logic/modules/auth/reducers.js]: LOGOUT_REQUEST

File: package.json [hunk 16]
```diff
@@ -108,5 +107,24 @@
   },
   "engines": {
     "npm": ">=2.0.0"
+  },
+  "jest": {
+    "rootDir": "",
+    "testPathDirs": ["client"],
```
Comment (rosendi) [inline package.json]: `<rootDir>/client`

File: package.json [hunk 17]
```diff
@@ -108,5 +107,24 @@
...
+      "node_modules",
+      "src",
+      "client"
+    ],
+    "moduleNameMapper": {
+      "^.+\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "identity-obj-proxy",
+      "^.+\\.(css|less)$": "identity-obj-proxy"
+    },
+    "collectCoverageFrom": [
+      "client/logic/**/*.js",
```
Comment (rosendi) [inline package.json]: client/**/*.js

File: package.json [hunk 18]
```diff
@@ -108,5 +107,24 @@
...
+    "rootDir": "",
+    "testPathDirs": ["client"],
+    "testResultsProcessor": "node_modules/jest-teamcity-reporter",
+    "moduleDirectories": [
+      "node_modules",
+      "src",
+      "client"
+    ],
+    "moduleNameMapper": {
+      "^.+\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "identity-obj-proxy",
```
Comment (rosendi) [inline package.json]: jpg|png|gif|svg|ttf|woff|css

File: package.json [hunk 19]
```diff
@@ -108,5 +107,24 @@
...
+      "^.+\\.(css|less)$": "identity-obj-proxy"
```
Comment (rosendi) [inline package.json]: remove

File: client/logic/modules/auth/__tests__/reducers.js [hunk 20]
```diff
@@ -0,0 +1,80 @@
+import authReducer, {
+  AUTHENTICATE,
+  FETCH_USERDATA_REQUEST,
+  FETCH_USERDATA_SUCCESS,
+  FETCH_USERDATA_FAILURE,
+  LOGOUT_SUCCESS,
+} from '../';
+
```
Comment (rosendi) [inline client/logic/modules/auth/__tests__/reducers.js]: describe

File: client/lib/bem/__tests__/bem.js [hunk 21]
```diff
@@ -0,0 +1,25 @@
+import bem from '../';
+
+test('should return block className for :root', () => {
+  const className = bem('block')(':root').className;
+
+  expect(className).toBe('block');
+});
+
+test('should return element className for element', () => {
```
Comment (zouxuoz) [inline client/lib/bem/__tests__/bem.js]: fixed

File: client/lib/bem/__tests__/bem.js [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+  const className = bem('block')('element').className;
+
+  expect(className).toBe('block__element');
+});
+
+test('should return element className with boolean modifiers for element with boolean modifiers', () => {
```
Comment (zouxuoz) [inline client/lib/bem/__tests__/bem.js]: fixed

File: client/lib/bem/bem.js [hunk 23]
```diff
@@ -7,15 +7,19 @@ const bem = block => (element, modifiers = {}) => {
     rootClassName = `${block}__${element}`;
   }
 
-  const modifierClassNames = Object.keys(modifiers).map((modifier) => {
-    const value = modifiers[modifier];
+  const modifierClassNames = Object.keys(modifiers).reduce((result, modifierName) => {
+    const value = modifiers[modifierName];
 
     if (typeof value === 'boolean') {
```
Comment (zouxuoz) [inline client/lib/bem/bem.js]: fixed

File: client/logic/modules/auth/__tests__/creators.js [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+  LOGOUT_SUCCESS,
+  LOGOUT_FAILURE,
+  authenticate,
+  logout,
+  logoutDone,
+  logoutError,
+  fetchUserdata,
+  fetchUserdataDone,
+  fetchUserdataError,
+} from '../';
```
Comment (zouxuoz) [inline client/logic/modules/auth/__tests__/creators.js]: fixed

File: client/logic/modules/auth/__tests__/creators.js [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+test('should create an action for authenticate', () => {
+  const expectedAction = {
+    type: AUTHENTICATE,
+    payload: 'payload',
+  };
+
+  expect(authenticate('payload')).toEqual(expectedAction);
+});
+
+test('should create an action for logout', () => {
```
Comment (zouxuoz) [inline client/logic/modules/auth/__tests__/creators.js]: fixed

File: client/logic/modules/auth/__tests__/creators.js [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+  const expectedAction = {
+    type: LOGOUT_REQUEST,
+  };
+
+  expect(logout()).toEqual(expectedAction);
+});
+
+test('should create an action for logout done', () => {
```
Comment (zouxuoz) [inline client/logic/modules/auth/__tests__/creators.js]: fixed

File: client/logic/modules/auth/__tests__/creators.js [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+  const expectedAction = {
+    type: LOGOUT_SUCCESS,
+  };
+
+  expect(logoutDone()).toEqual(expectedAction);
+});
+
+test('should create an action for logout error', () => {
```
Comment (zouxuoz) [inline client/logic/modules/auth/__tests__/creators.js]: fixed

File: client/logic/modules/auth/__tests__/creators.js (hunk 11 above)
Comment (zouxuoz) [inline client/logic/modules/auth/__tests__/creators.js]: fixed

File: client/logic/modules/auth/__tests__/creators.js (hunk 13 above)
Comment (zouxuoz) [inline client/logic/modules/auth/__tests__/creators.js]: fixed

File: .gitignore (hunk 14 above)
Comment (zouxuoz) [inline .gitignore]: fixed

File: client/logic/modules/auth/reducers.js (hunk 15 above)
Comment (zouxuoz) [inline client/logic/modules/auth/reducers.js]: fixed

File: package.json (hunk 16 above)
Comment (zouxuoz) [inline package.json]: fixed

File: package.json (hunk 19 above)
Comment (zouxuoz) [inline package.json]: fixed

File: package.json (hunk 17 above)
Comment (zouxuoz) [inline package.json]: fixed

File: .gitignore [hunk 28]
```diff
@@ -1,6 +1,6 @@
 *.iml
-.DS_Store
 .babelcache
+.DS_Store
```
Comment (rosendi) [inline .gitignore]: ???

File: client/lib/bem/__tests__/bem.js [hunk 29]
```diff
@@ -0,0 +1,33 @@
+import bem from '../';
+
+describe('bem', () => {
+  let classes;
+
+  beforeEach(() => {
+    classes = bem('block');
+  });
+
+  test('returns a block`s class name for :root', () => {
```
Comment (rosendi) [inline client/lib/bem/__tests__/bem.js]: ’ that one

File: client/lib/bem/__tests__/bem.js [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+    const className = classes(':root').className;
+
+    expect(className).toBe('block');
+  });
+
+  test('returns an element`s class name for element', () => {
```
Comment (rosendi) [inline client/lib/bem/__tests__/bem.js]: ’ that one

File: .gitignore (hunk 28 above)
Comment (zouxuoz) [inline .gitignore]: it's ok

File: client/lib/bem/__tests__/bem.js (hunk 30 above)
Comment (zouxuoz) [inline client/lib/bem/__tests__/bem.js]: fixed 🐙

**Pull request merged**